reactor-core/src/main/java/reactor/util/observability/SequenceListenerFactory.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2022 VMware Inc. or its affiliates, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.util.observability;
+
+import org.reactivestreams.Publisher;
+
+import reactor.util.context.ContextView;
+
+/**
+ * A factory for per-subscription {@link SequenceListener}, exposing the ability to generate common state at publisher level
+ * from the source {@link Publisher}.
+ * <p>
+ * Examples of such state include:
+ * <ul>
+ *     <li>is the publisher a Mono? (unlocking Mono-specific behavior in the {@link SequenceListener}</li>
+ *     <li>resolution of NAME and TAGS on the source</li>
+ *     <li>preparation of a common configuration, like a registry for metrics</li>
+ * </ul>
+ *
+ * @param <T> the type of data emitted by the observed source {@link Publisher}
+ * @param <STATE> the type of the publisher-level state that will be shared between all {@link SequenceListener} created by this factory
+ * @author Simon Baslé
+ */
+public interface SequenceListenerFactory<T, STATE> {
```

Comment:
the public API can and does have a simplified version `Flux.listen(Supplier)`, but I do strongly believe the STATE is an important feature. Passing a `Scannable` is too likely to run into `Scannable.SCAN_UNAVAILABLE` instances.

This approach enables calls like `flux.listen(Micrometer.metrics())` with powerful configurability (the metrics can use a app-wide configuration for the `MeterRegistry` + discover name and tags on the source `flux` or its parents)